pkg/leader/leader.go

```diff
@@ -0,0 +1,194 @@
+// Copyright 2018 The Operator-SDK Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package leader
+
+import (
+	"context"
+	"errors"
+	"io/ioutil"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/rest"
+	crclient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// ErrNoNS indicates that a namespace could not be found for the current
+// environment
+var ErrNoNS = errors.New("namespace not found for current environment")
+
+// maxBackoffInterval defines the maximum amount of time to wait between
+// attempts to become the leader.
+const maxBackoffInterval = time.Second * 16
+
+// TryBecome behaves like Become, except it will not return an error in the
+// case where a namespace cannot be found for the current pod. This is useful
+// for a service that might run outside the cluster, for example an operator
+// being started with `operator-sdk up local`.
+func TryBecome(ctx context.Context, lockName string) error {
```

Comment:
For the `operator-sdk up local` case all we want to do is detect if the operator is not running in a pod and log that we're skipping leader election. We should do that in `Become()`. Otherwise users will have to keep changing their code between `Become()` and `TryBecome()` depending on how they run the operator.

So let's just remove `TryBecome()` and change the following in `Become()`:
```Go
ns, err := myNS()
if err != nil {
  if err == ErrNoNS {
    logrus.Info("Skipping leader election; not running in cluster")
    return nil
  }
  return err
}
```

Reply:
Good point. I'm not quite sure what you mean by needing to "keep changing their code...", but I did write this for two uses cases, and you're right that the SDK will probably only care about one of them. Just for the sake of discussion, they are:

1. My service might run off-cluster, so I want failure to detect the current namespace to be accepted as normal.
2. My service always runs in-cluster, so failure to detect the current namespace should result in hard failure.

For our purposes, I'll go ahead with the change you suggested. Thanks!